### PR TITLE
feat: add endpoint to create product

### PR DESCRIPTION
**Pull request by Juninhotech**
# Description
An endpoint to create new product

**Link to issue here**
#84 

# Changes proposed

## What were you told to do?

- Create a backend endpoint to handle product creation requests.
- Validate input data to ensure all necessary fields are provided and correctly formatted.
- Ensure proper error handling and response statuses.
- Update any related data or references accordingly.

## What did you do?
1. Implemented an endpoint to create new product
2.  created new dto for product creation

# Check List (Check all the applicable boxes)

🚨Please review the [contribution guideline](CONTRIBUTING.md) for this repository.

- [x] My code follows the code style of this project.
- [x] This PR does not contain plagiarized content.
- [x] The title and description of the PR is clear and explains the approach.
- [x] I am making a pull request against the **dev branch** (left side).
- [x] My commit messages styles matches our requested structure.
- [x] My code additions will fail neither code linting checks nor unit test.
- [x] I am only making changes to files I was requested to.

# Screenshots/Videos
image

### CODE REVIEW

**Comment by Juninhotech** [comment]
@Ace2489,  @chimaihueze

File: src/Hng.Application.Test/Features/Products/CreateProductTest.cs [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+{
+    public class CreateProductTest
+    {
+        private readonly Mock<IRepository<Product>> _mock;
+        private readonly Mock<IMapper> _mockMapper;
+        private readonly CreateProductHandler _handler;
+
+        public CreateProductTest()
+        {
+            _mock = new Mock<IRepository<Product>>();
```
Comment (chimaihueze) [inline src/Hng.Application.Test/Features/Products/CreateProductTest.cs:25]: You shouldn't be mocking Mapper. Look at the tests written by others

File: src/Hng.Application/Features/Products/Dtos/ProductCreationDto.cs [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+        public string Name { get; set; }
+
+        [JsonPropertyName("description")]
+        public string Description { get; set; }
+
+        [JsonPropertyName("category")]
+        public string Category { get; set; }
+
+        [Range(0.01, double.MaxValue, ErrorMessage = "Price must be a positive number")]
+
```
Comment (chimaihueze) [inline src/Hng.Application/Features/Products/Dtos/ProductCreationDto.cs:23]: why the extra space?

**Review by chimaihueze — CHANGES_REQUESTED** [review]

File: src/Hng.Application.Test/Features/Products/CreateProductTest.cs [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+            Assert.NotNull(result);
+            Assert.IsType<ProductDto>(result);
+            Assert.Equal(productDto.Id, result.Id);
+            Assert.Equal(productDto.Name, result.Name);
+            Assert.Equal(productDto.Description, result.Description);
+            Assert.Equal(productDto.Price, result.Price);
+
+            _mock.Verify(r => r.AddAsync(It.IsAny<Product>()), Times.Once);
+            _mockMapper.Verify(m => m.Map<Product>(It.IsAny<ProductCreationDto>()), Times.Once);
```
Comment (chimaihueze) [inline src/Hng.Application.Test/Features/Products/CreateProductTest.cs:72]: No need for the mapper here

**Review by chimaihueze — CHANGES_REQUESTED** [review]

File: src/Hng.Application/Features/Products/Dtos/ProductDto.cs [hunk 4]
```diff
@@ -1,9 +1,38 @@
...
+        [JsonPropertyName("description")]
+        public string Description { get; set; }
+
+        [JsonPropertyName("category")]
+        public string Category { get; set; }
+
+        public decimal Price { get; set; }
+
+        [JsonPropertyName("userid")]
+        public Guid UserId { get; set; }
```
Comment (NdubuisiJr) [inline src/Hng.Application/Features/Products/Dtos/ProductDto.cs:22]: This should be `user_id`

File: src/Hng.Application/Features/Products/Handlers/CreateProductHandler.cs [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    public class CreateProductHandler : IRequestHandler<CreateProductCommand, ProductDto>
+    {
+        private readonly IRepository<Product> _repository;
+        private readonly IMapper _mapper;
+        public CreateProductHandler(IRepository<Product> productReposiotry, IMapper mapper)
+        {
+            _repository = productReposiotry;
+            _mapper = mapper;
+        }
+        public async Task<ProductDto> Handle(CreateProductCommand request, CancellationToken cancellationToken)
```
Comment (NdubuisiJr) [inline src/Hng.Application/Features/Products/Handlers/CreateProductHandler.cs:24]: Add space between constructor and Handle method

File: src/Hng.Infrastructure/Context/ApplicationDbContext.cs [hunk 6]
```diff
@@ -9,6 +9,10 @@ protected override void OnModelCreating(ModelBuilder modelBuilder)
         {
             base.OnModelCreating(modelBuilder);
 
+            modelBuilder.Entity<Product>()
```
Comment (NdubuisiJr) [inline src/Hng.Infrastructure/Context/ApplicationDbContext.cs:12]: Take this out. This is the default

**Review by NdubuisiJr — CHANGES_REQUESTED** [review]

**Comment by Juninhotech** [comment]
@NdubuisiJr

**Pull request closed without merging**